From d6030db76fa0faecae51ac3b46526ec65dfa73cf Mon Sep 17 00:00:00 2001
From: Dinesh kumar <dineshkumar.varadarajan@adlinktech.com>
Date: Fri, 5 Apr 2019 18:55:29 +0530
Subject: [PATCH 31/47] Our SD/MMC interfaces are used differently.

- We can gate power to SD2, so add a regulator definition for that. The
  GPIO number (4,11) is the closest thing to an 'address' and avoids
  collisions without having to renumber anything.
- Quite different pinmux, of course.
- We use SD1,2,4 instead of interfaces 2,3,4.
- Changed feature sets (WP pin, change detect, data width, 1.8V support,
  suspend behavior).

Signed-off-by: Dinesh kumar <dineshkumar.varadarajan@adlinktech.com>
---
 arch/arm/boot/dts/lec-imx6.dtsi | 85 +++++++++++++++++++++++------------------
 1 file changed, 48 insertions(+), 37 deletions(-)

diff --git a/arch/arm/boot/dts/lec-imx6.dtsi b/arch/arm/boot/dts/lec-imx6.dtsi
index 15e96ca..f225872 100644
--- a/arch/arm/boot/dts/lec-imx6.dtsi
+++ b/arch/arm/boot/dts/lec-imx6.dtsi
@@ -117,6 +117,17 @@
 			vin-supply = <&reg_smarc_usb1>;
 		};
 
+		reg_smarc_sdio_pwr: regulator@411 {
+			compatible = "regulator-fixed";
+			reg = <411>;
+			regulator-name = "smarc_sdio_pwr";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			gpio = <&gpio4 11 GPIO_ACTIVE_HIGH>; /* SMARC SDIO_PWR_EN */
+			enable-active-high;
+			regulator-boot-on;
+		};
+
 		reg_audio: regulator@2 {
 			compatible = "regulator-fixed";
 			reg = <2>;
@@ -1020,38 +1031,37 @@
 			>;
 		};
 
+		pinctrl_usdhc1: usdhc1grp {
+			fsl,pins = <	/* on-module eMMC */
+				MX6QDL_PAD_SD1_CMD__SD1_CMD		0x17059
+				MX6QDL_PAD_SD1_CLK__SD1_CLK		0x10059
+				MX6QDL_PAD_SD1_DAT0__SD1_DATA0		0x17059
+				MX6QDL_PAD_SD1_DAT1__SD1_DATA1		0x17059
+				MX6QDL_PAD_SD1_DAT2__SD1_DATA2		0x17059
+				MX6QDL_PAD_SD1_DAT3__SD1_DATA3		0x17059
+				MX6QDL_PAD_NANDF_D0__SD1_DATA4		0x17059
+				MX6QDL_PAD_NANDF_D1__SD1_DATA5		0x17059
+				MX6QDL_PAD_NANDF_D2__SD1_DATA6		0x17059
+				MX6QDL_PAD_NANDF_D3__SD1_DATA7		0x17059
+			>;
+		};
+
 		pinctrl_usdhc2: usdhc2grp {
-			fsl,pins = <
+			fsl,pins = <	/* SMARC SDIO */
 				MX6QDL_PAD_SD2_CMD__SD2_CMD		0x17059
 				MX6QDL_PAD_SD2_CLK__SD2_CLK		0x10059
 				MX6QDL_PAD_SD2_DAT0__SD2_DATA0		0x17059
 				MX6QDL_PAD_SD2_DAT1__SD2_DATA1		0x17059
 				MX6QDL_PAD_SD2_DAT2__SD2_DATA2		0x17059
 				MX6QDL_PAD_SD2_DAT3__SD2_DATA3		0x17059
-				MX6QDL_PAD_NANDF_D4__SD2_DATA4		0x17059
-				MX6QDL_PAD_NANDF_D5__SD2_DATA5		0x17059
-				MX6QDL_PAD_NANDF_D6__SD2_DATA6		0x17059
-				MX6QDL_PAD_NANDF_D7__SD2_DATA7		0x17059
-			>;
-		};
-
-		pinctrl_usdhc3: usdhc3grp {
-			fsl,pins = <
-				MX6QDL_PAD_SD3_CMD__SD3_CMD		0x17059
-				MX6QDL_PAD_SD3_CLK__SD3_CLK		0x10059
-				MX6QDL_PAD_SD3_DAT0__SD3_DATA0		0x17059
-				MX6QDL_PAD_SD3_DAT1__SD3_DATA1		0x17059
-				MX6QDL_PAD_SD3_DAT2__SD3_DATA2		0x17059
-				MX6QDL_PAD_SD3_DAT3__SD3_DATA3		0x17059
-				MX6QDL_PAD_SD3_DAT4__SD3_DATA4		0x17059
-				MX6QDL_PAD_SD3_DAT5__SD3_DATA5		0x17059
-				MX6QDL_PAD_SD3_DAT6__SD3_DATA6		0x17059
-				MX6QDL_PAD_SD3_DAT7__SD3_DATA7		0x17059
+				//MX6QDL_PAD_GPIO_4__SD2_CD_B	     0x80000000 /* controller internal CD doesn't work */
+				MX6QDL_PAD_GPIO_4__GPIO1_IO04	     0x80000000 /* SD2_CD_B: mux as GPIO instead */
+				MX6QDL_PAD_GPIO_2__SD2_WP	     0x80000000
 			>;
 		};
 
 		pinctrl_usdhc4: usdhc4grp {
-			fsl,pins = <
+			fsl,pins = <	/* SMARC SDMMC */
 				MX6QDL_PAD_SD4_CMD__SD4_CMD		0x17059
 				MX6QDL_PAD_SD4_CLK__SD4_CLK		0x10059
 				MX6QDL_PAD_SD4_DAT0__SD4_DATA0		0x17059
@@ -1062,6 +1072,7 @@
 				MX6QDL_PAD_SD4_DAT5__SD4_DATA5		0x17059
 				MX6QDL_PAD_SD4_DAT6__SD4_DATA6		0x17059
 				MX6QDL_PAD_SD4_DAT7__SD4_DATA7		0x17059
+				MX6QDL_PAD_NANDF_ALE__SD4_RESET	     0x80000000
 			>;
 		};
 
@@ -1199,37 +1210,37 @@
 	fsl,tx-d-cal = <106>;
 };
 
-&usdhc2 {
+&usdhc1 {	/* on-module eMMC */
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_usdhc2>;
+	pinctrl-0 = <&pinctrl_usdhc1>;
 	bus-width = <8>;
-	cd-gpios = <&gpio2 2 GPIO_ACTIVE_LOW>;
-	wp-gpios = <&gpio2 3 GPIO_ACTIVE_HIGH>;
+	non-removable;
 	no-1-8-v;
-	keep-power-in-suspend;
-	enable-sdio-wakeup;
 	status = "okay";
 };
 
-&usdhc3 {
+&usdhc2 {	/* SMARC SDIO */
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_usdhc3>;
-	bus-width = <8>;
-	cd-gpios = <&gpio2 0 GPIO_ACTIVE_LOW>;
-	wp-gpios = <&gpio2 1 GPIO_ACTIVE_HIGH>;
+	pinctrl-0 = <&pinctrl_usdhc2>;
+	bus-width = <4>;
+	//fsl,cd-controller; /* using controller internal CD doesn't work properly */
+	cd-gpios = <&gpio1 4 GPIO_ACTIVE_LOW>; /* need to treat it as GPIO instead */
+	fsl,wp-controller;
+	vmmc-supply = <&reg_smarc_sdio_pwr>;
+	vqmmc-supply = <&reg_smarc_sdio_pwr>;
 	no-1-8-v;
-	keep-power-in-suspend;
-	enable-sdio-wakeup;
+	//keep-power-in-suspend;
+	//enable-sdio-wakeup;
 	status = "okay";
 };
 
-&usdhc4 {
+&usdhc4 {	/* SMARC SDMMC */
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_usdhc4>;
 	bus-width = <8>;
-	non-removable;
+	//non-removable; /* eMMC: no polling saves power */
+	broken-cd; /* uSD slot on carrier: polling allows switching cards */
 	no-1-8-v;
-	keep-power-in-suspend;
 	status = "okay";
 };
 
-- 
2.7.4

